sdk/python/kfp/_credentials.py

```diff
@@ -0,0 +1,37 @@
+# Copyright 2021 Arrikto Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+
+__all__ = [
+    "TokenCredentials",
+]
+
+
+class TokenCredentials(object):
+
+    __metaclass__ = abc.ABCMeta
+
+    def refresh_api_key_hook(self, config):
+        """Refresh the api key.
+
+        This is a helper function for registering token refresh with swagger
+        generated clients.
+        """
+        config.api_key["authorization"] = self.get_token()
+
+    @abc.abstractmethod
+    def get_token(self):
+        raise NotImplementedError()
```

Comment:
Because we refresh api_key every time before sending a request, I feel like this method in interface is not necessary. It's enough having just refresh_api_key_hook.

Reply:
We might want to rename that method, what do you think about "update_auth_token(request_config)"

Reply:
@Bobgy 

>It's enough having just refresh_api_key_hook.

I think removing the `get_token()` method [in favor of a different unique method which unifies functionalities] is error prone. 
We don't want users to be messing with the `refresh_api_key_hook()` method (unless they have a strong reason to do so). What it does (setting the key for the `authorization` header) is very specific and important for the functionality.

Users should only be declaring the logic with which the client can retrieve[/get/read/generate] the token. The rest is implementation detail (and following the codebase as described in [my issue comment](https://github.com/kubeflow/pipelines/issues/5138#issuecomment-797003710), the K8s client defines it strictly enough).

Plus, we have thought that this name is a good choice because the `Configuation` object calls it that way itself (`configuration.refresh_api_key_hook`).

Am I explaining it properly? Do the above make any sense?

Reply:
> We don't want users to be messing with the refresh_api_key_hook() method (unless they have a strong reason to do so). What it does (setting the key for the authorization header) is very specific and important for the functionality.

I'm not fully confident to say so, e.g. https://www.kubeflow.org/docs/distributions/aws/pipeline/#authenticate-kubeflow-pipeline-using-sdk-inside-cluster needs a cookie to authenticate. Should that be avoided or sth potentially needed for other scenarios as well?

Rethinking about this, providing a default refresh_api_key_hook implementation while allowing people to override seems reasonable too, so I think I'm OK with this now.

The configuration object is auto-generated code, so we do not have a naming choice. We might consider change the auto-generator, so it's not necessarily a name we need to stick to.
Anyway, I feel like it's not a bad choice either.